### DISCUSSION TITLE
How to convert yolov5x (and v5x6) to onnx?

### DISCUSSION THREAD

**Discussion by NMZ0429** [zhiqwang/yolov5-rt-stack#279] [Q&A] [answered]
First of all, thank you so much for this wonderful library.

As I followed the example at `notebooks/export-onnx-inference-onnxruntime.ipynb`, the conversion to onnx was successful. However, large yolov5 such as yolov5x hasn't been implemented according to `models/__init_.py`[here](https://github.com/zhiqwang/yolov5-rt-stack/blob/3485ea144fec7b2857d0d2e0d4ff329959e77027/yolort/models/__init__.py#L11-L22) . 

How would  it be possible to convert unimplemented models to onnx? I saw someone converted it in the discussion  [https://github.com/zhiqwang/yolov5-rt-stack/discussions/160#discussion-3557543](https://github.com/zhiqwang/yolov5-rt-stack/discussions/160#discussion-3557543)

**Comment by zhiqwang** [comment, marked as answer]
Hi @NMZ0429 , Thanks for adding this ticket.

We've added the support for yolov5x and yolov5x6 in #261 . I haven't added its definition in `models/__init__.py` since we haven't uploaded the translated model yet. (We'll upload the translated PyTorch checkpoint at this week.) But you could find these in `models/yolo.py`:

https://github.com/zhiqwang/yolov5-rt-stack/blob/3485ea144fec7b2857d0d2e0d4ff329959e77027/yolort/models/yolo.py#L31-L32

And you can now use the following ways to load `yolov5x6` in `notebooks/export-onnx-inference-onnxruntime.ipynb`:

```python
from yolort.models import YOLOv5

# 'yolov5x.pt' is downloaded from https://github.com/ultralytics/yolov5/releases/download/v6.0/yolov5x6.pt
ckpt_path_from_ultralytics = "yolov5x6.pt"
model = YOLOv5.load_from_yolov5(ckpt_path_from_ultralytics, score_thresh=0.25)

model.eval()
img_path = "test/assets/bus.jpg"
predictions = model.predict(img_path)
```

Besides, we provide a CLI tool to export the ONNX model trained from ultralytics/yolov5 in https://github.com/zhiqwang/yolov5-rt-stack/blob/main/tools/export_model.py, I guess you could use the following script to export the ONNX model of `yolov5x6.pt`

```
python tools/export_model.py --checkpoint_path "./yolov5x6.pt"
```

**Reply by anshudaur** [reply]
Hello @zhiqwang ,

Thank you for your work.
I am trying out to export custom 5x6 model to onnx format :
But i am running to opset version errors:

RuntimeError: Exporting the operator silu to ONNX opset version 11 is not supported. Please open a bug to request ONNX export support for the missing operator.
The imported opset version from torchvison is 11.

Can you please tell me what is wrong here?

Thank you
Anshu

**Reply by zhiqwang** [reply]
Hi @anshudaur , 

I guess you're using PyTorch 1.7, you need to wrap the model by `_export_module_friendly(model)` as we do in 

https://github.com/zhiqwang/yolov5-rt-stack/blob/3485ea144fec7b2857d0d2e0d4ff329959e77027/yolort/models/__init__.py#L39

https://github.com/zhiqwang/yolov5-rt-stack/blob/3485ea144fec7b2857d0d2e0d4ff329959e77027/yolort/models/__init__.py#L189-L196

PyTorch supports exporting `nn.SiLU` to ONNX from 1.8 onwards, and we suggest you upgrade your PT to 1.9 or above for easier use.

**Reply by anshudaur** [reply]
Got it, thanks.
